lib/linter/apply-disable-directives.js

```diff
@@ -197,62 +197,99 @@ function processUnusedDisableDirectives(allDirectives) {
-                    if (directive.ruleId === null) {
-                        currentGlobalDisableDirective = null;
-                        disabledRuleMap.clear();
-                    } else if (currentGlobalDisableDirective) {
-                        enabledRules.add(directive.ruleId);
-                        disabledRuleMap.delete(directive.ruleId);
-                    } else {
-                        disabledRuleMap.delete(directive.ruleId);
+                    if (directive.ruleId === null || directive.ruleId === problem.ruleId) {
+                        if (
+                            problem.suppressions && (
+                                directive.unprocessedDirective.type === "disable-line" ||
+                                directive.unprocessedDirective.type === "disable-next-line"
+                            )
+                        ) {
+                            problem.suppressions.pop();
+                        } else {
+                            problem.suppressions = [];
+                        }
+                    }
+
+                    if (!isProblemSuppressedBefore) {
+                        if (directive.ruleId === null || currentGlobalDisableDirective) {
+                            currentGlobalDisableDirective = null;
+                        }
+                        if (directive.ruleId === null || directive.ruleId === problem.ruleId) {
+                            usedDisableDirectivesForProblem.forEach(usedDirective => {
+                                if (
+                                    usedDirective.ruleId === problem.ruleId &&
+                                    (!problem.suppressions || problem.suppressions.length === 0)
+                                ) {
+                                    usedDisableDirectivesForProblem.delete(usedDirective);
+                                }
+                            });
+                        }
                     }
                     break;
 
                 // no default
             }
         }
 
-        if (disabledRuleMap.has(problem.ruleId)) {
-            usedDisableDirectives.add(disabledRuleMap.get(problem.ruleId));
-        } else if (currentGlobalDisableDirective && !enabledRules.has(problem.ruleId)) {
-            usedDisableDirectives.add(currentGlobalDisableDirective);
-        } else {
-            problems.push(problem);
+        if (currentGlobalDisableDirective) {
+            const usedDisableDirectiveWithRuleId = [...usedDisableDirectivesForProblem].filter(directive => directive.ruleId === problem.ruleId);
+
+            if (usedDisableDirectiveWithRuleId) {
+                usedDisableDirectivesForProblem.delete(usedDisableDirectiveWithRuleId[0]);
+            }
+            usedDisableDirectivesForProblem.add(currentGlobalDisableDirective);
+        }
+        usedDisableDirectivesForProblem.forEach(usedDirective => {
+            usedDisableDirectives.add(usedDirective);
+        });
```

Comment:
Only one directive is considered as used for a problem. For example, in the following code one of these two directives should be reported as unused:

```js
/* eslint no-undef: 2 */

// eslint-disable-next-line no-undef
foo;  // eslint-disable-line no-undef

```

`main` branch:

```
D:\projects\eslint\foo.js
  3:1  error  Unused eslint-disable directive (no problems were reported from 'no-undef')

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.


D:\projects\eslint>
```

this branch:

```
D:\tmp\suppressions\eslint>npx eslint foo.js --report-unused-disable-directives

D:\tmp\suppressions\eslint>
```

The behavior regarding unused directives shouldn't be changed.

Reply:
Fixed and added more unit tests to cover these cases.